packages/fxa-auth-server/fxa-oauth-server/lib/jwt-token.js

```diff
@@ -0,0 +1,88 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+const hex = require('buf').to.hex;
+const AppError = require('./error');
+const JWTool = require('fxa-jwtool');
+
+const keys = require('../lib/keys');
+
+/**
+ * Create a JWT access token from `grant`
+ */
+exports.create = async function generateJWTAccessToken(db, grant) {
+  const accessToken = await db.generateAccessToken(grant);
+  const clientId = hex(grant.clientId);
+
+  // Claims list from:
+  // https://tools.ietf.org/html/draft-bertocci-oauth-access-token-jwt-00#section-2.2
+  const claims = {
+    // The IETF spec for `aud` refers to https://openid.net/specs/openid-connect-core-1_0.html#IDToken
+    // > REQUIRED. Audience(s) that this ID Token is intended for. It MUST contain the
+    // > OAuth 2.0 client_id of the Relying Party as an audience value. It MAY also contain
+    // > identifiers for other audiences. In the general case, the aud value is an array of
+    // > case sensitive strings. In the common special case when there is one audience, the
+    // > aud value MAY be a single case sensitive string.
+
+    // TODO: accept a "resource" parameter to endpoints where access tokens are returned.
+    // Convert "aud" to an array
+    aud: clientId,
+    client_id: clientId,
+    exp: Math.floor(accessToken.expiresAt / 1000),
+    iat: Math.floor(Date.now() / 1000),
+    iss: keys.ISSUER,
+    jti: hex(accessToken.token),
+    // scope is not in the spec but added so that the service provider can
+    // determine whether the presenter should be granted access to the resource
+    scope: grant.scope.toString(),
+    sub: hex(grant.userId),
+  };
+
+  return {
+    ...accessToken,
+    jwt_token: await keys.SIGNING_KEY.sign(claims),
+  };
+};
+
+/**
+ * Get the token ID of the JWT access token.
+ *
+ * @throws `invalidToken` error if access token is invalid.
+ */
+exports.tokenId = async function tokenId(accessToken) {
+  // The access token ID is stored in the jti field of
+  // a JWT access token.
+  const jwt = await exports.verify(accessToken);
+  if (! jwt.jti) {
+    throw AppError.invalidToken();
+  }
+  return jwt.jti;
+};
+
+/**
+ * Verify the form of a JWT access token.
+ *
+ * @returns {Object} JWT payload
+ * @throws `invalidToken` error if access token is invalid.
+ */
+exports.verify = async function verify(accessToken) {
+  let header;
+
+  try {
+    header = JWTool.unverify(accessToken).header;
+  } catch (e) {
+    throw AppError.invalidToken();
+  }
+  const signingKey = keys.publicJWK(header.kid);
+  // signingKey is undefined if header does not contain `kid`
+  if (!signingKey) {
+    throw AppError.invalidToken();
+  }
+  const payload = signingKey.verifySync(accessToken);
+  if (!payload) {
+    throw AppError.invalidToken();
+  }
+  return payload;
```

Comment:
There are additional things we should check here, to set a good example for other consumers of these tokens.  Things that might affect security include:

* The `typ` header must be present and exactly equal `"at+jwt"` (might need to fix fxa-jwtool to return this value)

Things that would be good hygiene but *shouldn't* ever happen in practice (because we know that it's signed with one of our keys, and we'd never sign invalid data) include:

* The `iss` claim should be present and exactly match our issuer value.
* Other expected claims should be present and validate against expected schema.

I suggest validating the claims against an explicit schema, like we do for browserid assertions [here](https://github.com/mozilla/fxa/blob/master/packages/fxa-auth-server/fxa-oauth-server/lib/assertion.js#L34).